src/main/java/org/jenkinsci/plugins/workflow/cps/EnvActionImpl.java

```diff
@@ -199,4 +203,29 @@ public Collection<EnvironmentContributor> getEnvironmentContributors() {
 
     }
 
+    @Extension public static class EnvActionImplPickleFactory extends SingleTypedPickleFactory<EnvActionImpl> {
+        @Override
+        protected Pickle pickle(EnvActionImpl object) {
+            return new EnvActionImplPickle();
+        }
+    }
+
+    /**
+     * Prevents multiple instances of {@link EnvActionImpl} from existing for a single Pipeline after a Jenkins restart
+     * in case {@code env} is serialized into the program.
+     */
+    private static class EnvActionImplPickle extends Pickle {
```

Comment:
This has no internal state, we just use it as a hook to return `EnvActionImpl.forRun(owner.getExecutable())` during deserialization, completely discarding the serialized object.